File: 04-Socket-UDP-Fondamenti/esempi/EchoUDP.java
/**
 * Nome dell'Esempio: Echo UDP Server e Client
 * Guida di Riferimento: 01-Socket-UDP-Base.md
 * 
 * Obiettivo: Dimostrare comunicazione UDP bidirezionale con gestione errori.
 * 
 * Spiegazione:
 * 1. Server UDP che rimanda indietro i messaggi ricevuti
 * 2. Client UDP che invia messaggi e riceve risposte
 * 3. Gestione timeout e pacchetti persi
 * 
 * @author Socket Programming Course
 * @version 1.0
 */

import java.io.*;
import java.net.*;

public class EchoUDP {
    
    /**
     * Server UDP Echo
     */
    public static class Server {
        private final int port;
        private DatagramSocket socket;
        private boolean running;
        
        public Server(int port) {
            this.port = port;
        }
        
        public void start() throws IOException {
            socket = new DatagramSocket(port);
            running = true;
            
            System.out.println("🚀 Server UDP Echo avviato su porta " + port);
            System.out.println("📦 Dimensione buffer: 1024 byte");
            System.out.println("🛑 Premi Ctrl+C per fermare");
            System.out.println("=" .repeat(50));
            
            byte[] buffer = new byte[1024];
            
            while (running) {
                try {
                    // Riceve pacchetto
                    DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                    socket.receive(packet);
                    
                    // Estrae informazioni
                    String received = new String(packet.getData(), 0, packet.getLength());
                    InetAddress clientAddress = packet.getAddress();
                    int clientPort = packet.getPort();
                    
                    System.out.println("📨 Ricevuto da " + clientAddress + ":" + clientPort + 
                                     " → " + received);
                    
                    // Prepara risposta (echo + timestamp)
                    String response = "ECHO: " + received + " [" + System.currentTimeMillis() + "]";
                    byte[] responseData = response.getBytes();
                    
                    // Invia risposta
                    DatagramPacket responsePacket = new DatagramPacket(
                        responseData, responseData.length, clientAddress, clientPort);
                    socket.send(responsePacket);
                    
                    System.out.println("📤 Risposta inviata: " + response);
                    
                } catch (IOException e) {
                    if (running) {
                        System.err.println("❌ Errore server: " + e.getMessage());
                    }
                }
            }
        }
        
        public void stop() {
            running = false;
            if (socket != null && !socket.isClosed()) {
                socket.close();
                System.out.println("🔒 Server fermato");
            }
        }
    }
    
    /**
     * Client UDP
     */
    public static class Client {
        private final String serverHost;
        private final int serverPort;
        private DatagramSocket socket;
        
        public Client(String serverHost, int serverPort) {
            this.serverHost = serverHost;
            this.serverPort = serverPort;
        }
        
        public String sendMessage(String message, int timeoutMs) throws IOException {
            if (socket == null) {
                socket = new DatagramSocket();
                socket.setSoTimeout(timeoutMs);
            }
            
            try {
                // Prepara e invia pacchetto
                byte[] data = message.getBytes();
                InetAddress address = InetAddress.getByName(serverHost);
                DatagramPacket packet = new DatagramPacket(data, data.length, address, serverPort);
                
                System.out.println("📤 Invio: " + message);
                socket.send(packet);
                
                // Riceve risposta
                byte[] buffer = new byte[1024];
                DatagramPacket responsePacket = new DatagramPacket(buffer, buffer.length);
                socket.receive(responsePacket);
                
                String response = new String(responsePacket.getData(), 0, responsePacket.getLength());
                System.out.println("📨 Risposta: " + response);
                
                return response;
                
            } catch (SocketTimeoutException e) {
                System.err.println("⏰ Timeout - nessuna risposta in " + timeoutMs + "ms");
                throw e;
            } catch (UnknownHostException e) {
                System.err.println("❌ Host sconosciuto: " + serverHost);
                throw e;
            }
        }
        
        public void close() {
            if (socket != null && !socket.isClosed()) {
                socket.close();
            }
        }
        
        public void interactive() {
            System.out.println("💬 Modalità interattiva avviata");
            System.out.println("💡 Digita 'quit' per uscire");
            System.out.println("-" .repeat(30));
            
            try (BufferedReader reader = new BufferedReader(new InputStreamReader(System.in))) {
                String input;
                int messageCount = 0;
                
                while (true) {
                    System.out.print("Tu: ");
                    input = reader.readLine();
                    
                    if (input == null || "quit".equalsIgnoreCase(input.trim())) {
                        break;
                    }
                    
                    if (input.trim().isEmpty()) {
                        continue;
                    }
                    
                    try {
                        messageCount++;
                        String messageWithId = "[#" + messageCount + "] " + input;
                        sendMessage(messageWithId, 5000);
                        
                    } catch (SocketTimeoutException e) {
                        System.err.println("⚠️ Messaggio perso (timeout)");
                    } catch (IOException e) {
                        System.err.println("❌ Errore invio: " + e.getMessage());
                    }
                    
                    System.out.println();
                }
                
                System.out.println("👋 Disconnessione...");
                
            } catch (IOException e) {
                System.err.println("❌ Errore input: " + e.getMessage());
            } finally {
                close();
            }
        }
    }
    
    public static void main(String[] args) {
        if (args.length == 0) {
            System.out.println("🛠️ Echo UDP - Server e Client");
            System.out.println("Utilizzo:");
            System.out.println("  java EchoUDP server <porta>");
            System.out.println("  java EchoUDP client <host> <porta>");
            System.out.println();
            System.out.println("Esempi:");
            System.out.println("  java EchoUDP server 9999");
            System.out.println("  java EchoUDP client localhost 9999");
            return;
        }
        
        String mode = args[0].toLowerCase();
        
        if ("server".equals(mode)) {
            if (args.length < 2) {
                System.err.println("❌ Porta richiesta per il server");
                return;
            }
            
            try {
                int port = Integer.parseInt(args[1]);
                Server server = new Server(port);
                
                // Shutdown hook per chiusura pulita
                Runtime.getRuntime().addShutdownHook(new Thread(server::stop));
                
                server.start();
                
            } catch (NumberFormatException e) {
                System.err.println("❌ Porta non valida: " + args[1]);
            } catch (IOException e) {
                System.err.println("💥 Errore avvio server: " + e.getMessage());
            }
            
        } else if ("client".equals(mode)) {
            if (args.length < 3) {
                System.err.println("❌ Host e porta richiesti per il client");
                return;
            }
            
            try {
                String host = args[1];
                int port = Integer.parseInt(args[2]);
                
                System.out.println("🔗 Client UDP Echo");
                System.out.println("Target: " + host + ":" + port);
                System.out.println("=" .repeat(30));
                
                Client client = new Client(host, port);
                
                // Test singolo messaggio
                try {
                    client.sendMessage("Test di connettività", 3000);
                    System.out.println("✅ Connessione OK\n");
                    
                    // Modalità interattiva
                    client.interactive();
                    
                } catch (IOException e) {
                    System.err.println("💥 Test connessione fallito: " + e.getMessage());
                }
                
            } catch (NumberFormatException e) {
                System.err.println("❌ Porta non valida: " + args[2]);
            }
            
        } else {
            System.err.println("❌ Modalità non riconosciuta: " + mode);
            System.out.println("Modalità disponibili: server, client");
        }
    }
}

File: 04-Socket-UDP/esempi/01.02-EchoUdp/echo-udp.js
const dgram = require('dgram');
const readline = require('readline');

/**
 * Nome dell'Esempio: Echo UDP Server e Client (JavaScript/Node.js)
 * Guida di Riferimento: 01-Socket-UDP-Base.md
 * 
 * Obiettivo: Dimostrare comunicazione UDP bidirezionale con gestione errori.
 * 
 * Spiegazione:
 * 1. Server UDP che rimanda indietro i messaggi ricevuti
 * 2. Client UDP che invia messaggi e riceve risposte
 * 3. Gestione timeout e pacchetti persi
 * 
 * @author Socket Programming Course
 * @version 1.0
 */

// ==================== SERVER UDP ECHO ====================

class EchoServer {
    constructor(port) {
        this.port = port;
        this.socket = null;
        this.running = false;
    }
    
    start() {
        this.socket = dgram.createSocket('udp4');
        this.running = true;
        
        this.socket.on('error', (err) => {
            if (this.running) {
                console.error(`❌ Errore server: ${err.message}`);
            }
            this.socket.close();
        });
        
        this.socket.on('listening', () => {
            const address = this.socket.address();
            console.log('🚀 Server UDP Echo avviato su porta ' + address.port);
            console.log('📦 Dimensione buffer: 1024 byte');
            console.log('🛑 Premi Ctrl+C per fermare');
            console.log('='.repeat(50));
        });
        
        this.socket.on('message', (msg, remote) => {
            const received = msg.toString();
            console.log(`📨 Ricevuto da ${remote.address}:${remote.port} → ${received}`);
            
            // Prepara risposta (echo + timestamp)
            const response = `ECHO: ${received} [${Date.now()}]`;
            const responseBuffer = Buffer.from(response);
            
            // Invia risposta
            this.socket.send(responseBuffer, 0, responseBuffer.length, remote.port, remote.address, (err) => {
                if (err) {
                    console.error(`❌ Errore invio risposta: ${err.message}`);
                } else {
                    console.log(`📤 Risposta inviata: ${response}`);
                }
            });
        });
        
        // Binding sulla porta
        this.socket.bind(this.port);
    }
    
    stop() {
        this.running = false;
        if (this.socket) {
            this.socket.close();
            console.log('🔒 Server fermato');
        }
    }
}

// ==================== CLIENT UDP ====================

class EchoClient {
    constructor(serverHost, serverPort) {
        this.serverHost = serverHost;
        this.serverPort = serverPort;
        this.socket = null;
    }
    
    sendMessage(message, timeoutMs) {
        return new Promise((resolve, reject) => {
            if (!this.socket) {
                this.socket = dgram.createSocket('udp4');
            }
            
            let timeoutId;
            let responseReceived = false;
            
            // Gestione timeout
            timeoutId = setTimeout(() => {
                if (!responseReceived) {
                    console.error(`⏰ Timeout - nessuna risposta in ${timeoutMs}ms`);
                    reject(new Error('Timeout'));
                }
            }, timeoutMs);
            
            // Listener per la risposta
            const onMessage = (msg, remote) => {
                if (!responseReceived) {
                    responseReceived = true;
                    clearTimeout(timeoutId);
                    
                    const response = msg.toString();
                    console.log(`📨 Risposta: ${response}`);
                    
                    resolve(response);
                }
            };
            
            // Registra listener
            this.socket.once('message', onMessage);
            
            // Gestione errori
            this.socket.once('error', (err) => {
                clearTimeout(timeoutId);
                console.error(`❌ Errore: ${err.message}`);
                reject(err);
            });
            
            // Prepara e invia pacchetto
            const buffer = Buffer.from(message);
            console.log(`📤 Invio: ${message}`);
            
            this.socket.send(buffer, 0, buffer.length, this.serverPort, this.serverHost, (err) => {
                if (err) {
                    clearTimeout(timeoutId);
                    console.error(`❌ Errore invio: ${err.message}`);
                    reject(err);
                }
            });
        });
    }
    
    close() {
        if (this.socket) {
            this.socket.close();
        }
    }
    
    async interactive() {
        console.log('💬 Modalità interattiva avviata');
        console.log('💡 Digita \'quit\' per uscire');
        console.log('-'.repeat(30));
        
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout,
            prompt: 'Tu: '
        });
        
        let messageCount = 0;
        
        rl.prompt();
        
        rl.on('line', async (input) => {
            const trimmed = input.trim();
            
            if (trimmed.toLowerCase() === 'quit') {
                console.log('👋 Disconnessione...');
                rl.close();
                this.close();
                process.exit(0);
                return;
            }
            
            if (trimmed === '') {
                rl.prompt();
                return;
            }
            
            try {
                messageCount++;
                const messageWithId = `[#${messageCount}] ${trimmed}`;
                await this.sendMessage(messageWithId, 5000);
            } catch (err) {
                if (err.message === 'Timeout') {
                    console.error('⚠️ Messaggio perso (timeout)');
                } else {
                    console.error(`❌ Errore invio: ${err.message}`);
                }
            }
            
            console.log();
            rl.prompt();
        });
        
        rl.on('close', () => {
            console.log('👋 Disconnessione...');
            this.close();
            process.exit(0);
        });
    }
}

// ==================== MAIN ====================

function printUsage() {
    console.log('🛠️ Echo UDP - Server e Client (Node.js)');
    console.log('Utilizzo:');
    console.log('  node echo-udp.js server <porta>');
    console.log('  node echo-udp.js client <host> <porta>');
    console.log();
    console.log('Esempi:');
    console.log('  node echo-udp.js server 9999');
    console.log('  node echo-udp.js client localhost 9999');
}

async function main() {
    const args = process.argv.slice(2);
    
    if (args.length === 0) {
        printUsage();
        process.exit(0);
    }
    
    const mode = args[0].toLowerCase();
    
    if (mode === 'server') {
        if (args.length < 2) {
            console.error('❌ Porta richiesta per il server');
            process.exit(1);
        }
        
        const port = parseInt(args[1]);
        
        if (isNaN(port)) {
            console.error(`❌ Porta non valida: ${args[1]}`);
            process.exit(1);
        }
        
        const server = new EchoServer(port);
        
        // Shutdown hook per chiusura pulita
        process.on('SIGINT', () => {
            console.log();
            server.stop();
            process.exit(0);
        });
        
        process.on('SIGTERM', () => {
            server.stop();
            process.exit(0);
        });
        
        server.start();
        
    } else if (mode === 'client') {
        if (args.length < 3) {
            console.error('❌ Host e porta richiesti per il client');
            process.exit(1);
        }
        
        const host = args[1];
        const port = parseInt(args[2]);
        
        if (isNaN(port)) {
            console.error(`❌ Porta non valida: ${args[2]}`);
            process.exit(1);
        }
        
        console.log('🔗 Client UDP Echo');
        console.log(`Target: ${host}:${port}`);
        console.log('='.repeat(30));
        
        const client = new EchoClient(host, port);
        
        // Test singolo messaggio
        try {
            await client.sendMessage('Test di connettività', 3000);
            console.log('✅ Connessione OK\n');
            
            // Modalità interattiva
            await client.interactive();
            
        } catch (err) {
            console.error(`💥 Test connessione fallito: ${err.message}`);
            client.close();
            process.exit(1);
        }
        
    } else {
        console.error(`❌ Modalità non riconosciuta: ${mode}`);
        console.log('Modalità disponibili: server, client');
        process.exit(1);
    }
}

// Avvia l'applicazione
main().catch((err) => {
    console.error('💥 Errore fatale:', err);
    process.exit(1);
});
